Jenkinsfile

```diff
@@ -1,20 +1,57 @@
 elifePipeline {
-    stage 'Checkout'
-    checkout scm
-    def commit = elifeGitRevision()
+    def commit
+    stage 'Checkout', {
+        checkout scm
+        commit = elifeGitRevision()
+    }
+
+    elifeOnNode(
+        {
+            stage 'Build image', {
+                checkout scm
+                sh "docker build -t elifesciences/api-dummy:${commit} ."
+            }
+        },
+        'elife-libraries--ci'
+    )
 
-    stage 'Project tests'
-    lock('api-dummy--ci') {
-        builderDeployRevision 'api-dummy--ci', commit
-        builderProjectTests 'api-dummy--ci', '/srv/api-dummy', ['/srv/api-dummy/build/phpunit.xml']
+    stage 'Project tests', {
+        lock('api-dummy--ci') {
+            builderDeployRevision 'api-dummy--ci', commit
+            builderProjectTests 'api-dummy--ci', '/srv/api-dummy', ['/srv/api-dummy/build/phpunit.xml']
+        }
     }
 
+    def image
+    elifeOnNode(
+        {
+            elifeMainlineOnly {
+                stage 'Push image', {
+                    checkout scm
+                    image = DockerImage.elifesciences('api-dummy', commit)
+                    image.push()
+                }
+            
+                stage 'Approval', {
+                    elifeGitMoveToBranch commit, 'approved'
+                    image.tag('approved').push()
+                }
+            }
+        },
+        'elife-libraries--ci'
+    )
+
     elifeMainlineOnly {
-        stage 'Approval'
-        elifeGitMoveToBranch commit, 'approved'
 
-        stage 'Deploy on demo'
-        elifeGitMoveToBranch commit, 'master'
-        builderDeployRevision 'api-dummy--demo', commit
+        stage 'Deploy on demo', {
+            elifeGitMoveToBranch commit, 'master'
+            elifeOnNode(
```

Comment:
the `elifeOnNode` are getting complex because we mix deploy to an EC2 instance with deploying an image - the target is to use only images and executes everything onto a node equipped for Docker/Kubernetes